Extend subgraph_rewriter logic

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51532 Extend subgraph_rewriter logic**

- Change output of `replace_pattern` to `List[Match]` reflecting the
pattern(s) matched in the original graph
- Ensure that all Callables (not just FunctionType objects) work with
the rewriter
- Fix incorrect matching in degenerate case (`test_subgraph_rewriter_correct_output_replacement`)
- Verify that pattern matching works when pattern and original graph are
the same

Differential Revision: [D26193082](https://our.internmc.facebook.com/intern/diff/D26193082)

Commit:
Extend subgraph_rewriter logic

- Change output of `replace_pattern` to `List[Match]` reflecting the
pattern(s) matched in the original graph
- Ensure that all Callables (not just FunctionType objects) work with
the rewriter
- Fix incorrect matching in degenerate case (`test_subgraph_rewriter_correct_output_replacement`)
- Verify that pattern matching works when pattern and original graph are
the same

[ghstack-poisoned]

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -60,25 +63,52 @@ def attributes_are_equal(pn : Node, gn : Node) -> bool:
         match_found = all(self._match_nodes(pn_, gn_) for pn_, gn_
                           in zip(pn.all_input_nodes, gn.all_input_nodes))
+        match_found = False
+        if pn.op == "output":
```
What I meant by the comment on the other PR was that going through permutations should be unnecessary so long as we are enforcing that the pattern graph only has one output value. In that case, I think this section should be something like:
```
match_found = any(self._match_nodes(pn_.all_input_nodes[0], gn_) for gn_ in gn.all_input_nodes)
```

jamesr66a replied:
Now that this is the logic, you'll want the fact that the pattern only has 1 output enforced. That's in the other PR here https://github.com/pytorch/pytorch/pull/51412/files#diff-14d3157db8fc26c5c1c0f18e780290a2b527c468dfc63cb3b64aa886c078502eR21 so if you land that one before this one (or put them in a stack) it should be fine

jamesr66a reviewed (commented):
Please address failing tests

Commit:
Update on "Extend subgraph_rewriter logic"


- Change output of `replace_pattern` to `List[Match]` reflecting the
pattern(s) matched in the original graph
- Ensure that all Callables (not just FunctionType objects) work with
the rewriter
- Fix incorrect matching in degenerate case (`test_subgraph_rewriter_correct_output_replacement`)
- Verify that pattern matching works when pattern and original graph are
the same

Differential Revision: [D26193082](https://our.internmc.facebook.com/intern/diff/D26193082)

[ghstack-poisoned]

Commit:
Update on "Extend subgraph_rewriter logic"


- Change output of `replace_pattern` to `List[Match]` reflecting the
pattern(s) matched in the original graph
- Ensure that all Callables (not just FunctionType objects) work with
the rewriter
- Fix incorrect matching in degenerate case (`test_subgraph_rewriter_correct_output_replacement`)
- Verify that pattern matching works when pattern and original graph are
the same

Differential Revision: [D26193082](https://our.internmc.facebook.com/intern/diff/D26193082)

[ghstack-poisoned]

jamesr66a reviewed (approved):
Awesome!

Commit:
Update on "Extend subgraph_rewriter logic"


- Change output of `replace_pattern` to `List[Match]` reflecting the
pattern(s) matched in the original graph
- Ensure that all Callables (not just FunctionType objects) work with
the rewriter
- Fix incorrect matching in degenerate case (`test_subgraph_rewriter_correct_output_replacement`)
- Verify that pattern matching works when pattern and original graph are
the same

Differential Revision: [D26193082](https://our.internmc.facebook.com/intern/diff/D26193082)

[ghstack-poisoned]
